### PR TITLE
Add guest properties support for `vm` and `vapp`

### PR DESCRIPTION
**Pull request by Didainius**
Both VMs and vApps allow to set guest properties which are useful for configuration (when guests support it).
This PR adds ~~`SetGuestProperties`~~ `SetProductSectionList` and ~~`GetGuestProperties`~~ `GetProductSectionList` for `vm` and `vapp` objects which allow to manipulate guest properties

### CODE REVIEW

File: govcd/guestproperties.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// setGuestProperties is a shared function for
+func setGuestProperties(client *Client, href string, properties *types.ProductSectionList) error {
+	if href == "" {
+		return fmt.Errorf("href cannot be empty to set guest properties")
+	}
+
+	properties.Xmlns = types.XMLNamespaceVCloud
+	properties.Ovf = types.XMLNamespaceOVF
+
+	task, err := client.ExecuteTaskRequest(href+"/productSections", http.MethodPut,
```
Comment (vbauzysvmware) [inline govcd/guestproperties.go]: kind'a lost here type producSections, but variable name is properties

File: govcd/guestproperties.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return fmt.Errorf("task for setting guest properties failed: %s", err)
+	}
+
+	return nil
+}
+
+// getGuestProperties is a shared function for both vApp and VM
+func getGuestProperties(client *Client, href string) (*types.ProductSectionList, error) {
```
Comment (vbauzysvmware) [inline govcd/guestproperties.go]: client receiver?

File: govcd/vapp_vm_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		},
+	}
+
+	gotProperties, err := object.SetGuestProperties(vappProperties)
+	check.Assert(err, IsNil)
+
+	getProperties, err := object.GetGuestProperties()
+	check.Assert(err, IsNil)
+
+	// Check that values were set in API
```
Comment (vbauzysvmware) [inline govcd/vapp_vm_test.go]: property size isn't checked

File: govcd/guestproperties.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+// setGuestProperties is a shared function for
+func setGuestProperties(client *Client, href string, properties *types.ProductSectionList) error {
```
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/guestproperties.go]: No need to pollute `client` struct for for a function usable only be two objects

File: govcd/guestproperties.go (hunk 2 above)
Comment (Didainius) [reply govcd/guestproperties.go]: No need to pollute `client` struct for for a function usable only be two objects

File: govcd/guestproperties.go (hunk 1 above)
Comment (Didainius) [reply govcd/guestproperties.go]: Yeah. The endpoint is such therefore struct name is as it is, but in fact we are dealing with "guest properties" from vCD perspective. It matches function names and UI.

File: govcd/vapp_vm_test.go (hunk 3 above)
Comment (Didainius) [reply govcd/vapp_vm_test.go]: true. fixed

File: govcd/guestproperties.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+// setGuestProperties is a shared function for
```
[the remark this answers is not in the document]
Comment (Didainius) [reply govcd/guestproperties.go]: fixed

**Comment by Didainius** [comment]
> Only question is how will the property updates be implemented on the Terraform side? Is the idea that an update of a single property in the list will cause the whole guest property struct to be re-set?

Yes. The idea is that whatever is set in `terraform` it will apply "as a whole" as in PR (https://github.com/terraform-providers/terraform-provider-vcd/pull/319) in line https://github.com/terraform-providers/terraform-provider-vcd/pull/319/files#diff-dad616393369ab476ae585e64e0c9e4dR1258

There is no mechanism to set "one by one" and even if there was - such method is way quicker than setting 100 properties in "one by one" fashion.

File: CHANGELOG.md [hunk 6]
```diff
@@ -22,6 +22,8 @@
 * Added methods `AdminOrg.GetVDCByName` and related `GetVDCById`, `GetVDCByNameOrId`
 * Added methods `AdminOrg.GetAdminVDCByName` and related `GetAdminVDCById`, `GetAdminVDCByNameOrId`
 * Added methods `Catalog.Refresh` and `AdminCatalog.Refresh`
+* Added methods `vm.SetGuestProperties` and `vm.GetGuestProperties` [#235](https://github.com/vmware/go-vcloud-director/pull/235)
+* Added methods `vapp.SetGuestProperties` and `vapp.GetGuestProperties` [#235](https://github.com/vmware/go-vcloud-director/pull/235)
```
Comment (lvirbalas) [inline CHANGELOG.md]: Structure has been renamed, but change log entry left old.

Comment (Didainius) [reply CHANGELOG.md]: I knew I had to miss it somewhere while renaming. Fixed.

File: govcd/vapp.go [hunk 7]
```diff
@@ -830,7 +833,26 @@ func updateNetworkConfigurations(vapp *VApp, networkConfigurations []types.VAppN
...
-// Function RemoveAllNetworks unattach all networks from VAPP
+// RemoveAllNetworks detaches all networks from vApp
 func (vapp *VApp) RemoveAllNetworks() (Task, error) {
 	return updateNetworkConfigurations(vapp, []types.VAppNetworkConfiguration{})
 }
+
+// SetProductSectionList sets product section for a vApp
+//
+// The slice of properties "ProductSectionList.ProductSection.Property" is not necessarily ordered or returned as set before
+func (vapp *VApp) SetProductSectionList(productSection *types.ProductSectionList) (*types.ProductSectionList, error) {
```
Comment (lvirbalas) [inline govcd/vapp.go]: I would add a sentence - a hint - that this product section represents guest properties in all comments of the four public functions.

Comment (Didainius) [reply govcd/vapp.go]: Added

Comment (lvirbalas) [reply govcd/vapp.go]: Typo in "maniapulate", sounds funny though :D

Comment (Didainius) [reply govcd/vapp.go]: fixed

**Pull request merged**